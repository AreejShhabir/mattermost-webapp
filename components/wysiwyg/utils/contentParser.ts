// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {JSONContent} from '@tiptap/react';
import {isNumber} from 'lodash';

/**
 * parse a JSONContent array and return a string containing markdown syntax
 */
const parseContentArray = (contentArr: JSONContent[], prefix: string|number = '', isTable = false): string => {
    let output = '';

    function applySingleMark(text: string, type: string, attrs?: Record<string, any>): string {
        switch (type) {
        case 'italic':
            return `*${text}*`;
        case 'bold':
            return `**${text}**`;
        case 'strike':
            return `~~${text}~~`;
        case 'code':
            return `\`${text}\``;
        case 'mmLink':
            return attrs?.href ? `[${text}](${attrs.href})` : text;
        default:
            return text;
        }
    }

    contentArr.forEach((content, index) => {
        /**
         * if the content is pure text wth marks applied to it (bold, italic, etc.) parse it here and return early,
         * since text-content cannot have children (content.content) anymore
         */
        if (content.type === 'text') {
            let {text = ''} = content;
            const {marks} = content;

            /**
             * test if we have any text at all or the text consists only of whitespace characters and simply
             * return it without formatting. Otherwise it could lead to strange combinations of MD-formatting chars
             */
            if (!text || (/^\s*$/g).test(text)) {
                return;
            }

            // apply the marks in order of appearance
            marks?.forEach(({type, attrs}) => {
                text = applySingleMark(text, type, attrs);
            });

            output += text;
        }

        if (!content.content) {
            return;
        }

        let value;
        const isLastItem = index === contentArr.length - 1;
        const lineBreak = isTable ? '' : '\n';

        switch (content.type) {
        case 'heading':
            output += `${prefix}${'#'.repeat(content.attrs?.level ?? 1)} ${parseContentArray(content.content, '', isTable)}${lineBreak.repeat(2)}`;
            return;
        case 'codeBlock':
            value = parseContentArray(content.content, '', isTable);
            if (isTable) {
                output += value;
                return;
            }
            output += `\`\`\`${content.attrs?.language ?? ''}\n${value}\n\`\`\`\n\n`;
            return;
        case 'blockquote':
            output += parseContentArray(content.content, '> ', isTable);
            return;
        case 'bulletList':
            output += parseContentArray(content.content, '- ', isTable);
            return;
        case 'orderedList':
            output += parseContentArray(content.content, content.attrs?.start ?? 1, isTable);
            return;
        case 'listItem':
            output += parseContentArray(content.content, isNumber(prefix) ? `${prefix + index}. ` : prefix, isTable);
            if (isLastItem) {
                output += lineBreak.repeat(2);
            }
            return;
        case 'table':
            output += parseContentArray(content.content, '', true);
            return;
        case 'tableRow':
            output += `| ${parseContentArray(content.content, '', isTable)}\n`;
            return;
        case 'tableHeader':
        case 'tableCell':
            output += `${parseContentArray(content.content, '', isTable)} |`;
            if (content.type === 'tableHeader' && isLastItem) {
                output += `\n${'|---'.repeat(contentArr.length)}|`;
            }
            return;
        case 'paragraph':
        default:
            value = parseContentArray(content.content, '', isTable);
            output += `${prefix}${value}${lineBreak}`;
        }
    });

    return output;
};

/**
 * Helper function to convert the JSONcontent AST syntax to Markdown.
 */
export function contentToMarkdown(content: JSONContent): string {
    if (!content.content) {
        return '';
    }

    return parseContentArray(content.content).trim();
}
