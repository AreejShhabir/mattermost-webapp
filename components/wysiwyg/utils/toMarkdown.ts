// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {fromHtml} from 'hast-util-from-html';
import {toMdast} from 'hast-util-to-mdast';
import {toMarkdown} from 'mdast-util-to-markdown';

// extension for github flavored markdown
import {gfmToMarkdown} from 'mdast-util-gfm';

import {unified} from 'unified';
import remarkParse from 'remark-parse';
import remarkGfm from 'remark-gfm';
import remarkRehype from 'remark-rehype';
import rehypeStringify from 'rehype-stringify';

const markdownToHtml = (markdown: string) => unified().
    use(remarkParse).
    use(remarkGfm).
    use(remarkRehype).
    use(rehypeStringify).
    processSync(markdown);

const htmlToMarkdown = (html: string) => {
    return toMarkdown(toMdast(fromHtml(html)), {extensions: [gfmToMarkdown()]});
};

export {htmlToMarkdown, markdownToHtml};
