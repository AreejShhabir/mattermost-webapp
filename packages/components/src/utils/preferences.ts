// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export function getPreferenceKey(category: string, name: string): string {
    return `${category}--${name}`;
}
