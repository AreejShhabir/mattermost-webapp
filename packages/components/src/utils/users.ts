// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {UserProfile} from '@mattermost/types/users';

import Preferences from '../constants/preferences';

import {localizeMessage} from './i18n';

export function getFullName(user: UserProfile): string {
    if (user.first_name && user.last_name) {
        return user.first_name + ' ' + user.last_name;
    } else if (user.first_name) {
        return user.first_name;
    } else if (user.last_name) {
        return user.last_name;
    }

    return '';
}

export function displayUsername(
    user: UserProfile | null | undefined,
    teammateNameDisplay: string,
    useFallbackUsername = true,
): string {
    let name = useFallbackUsername ? localizeMessage('channel_loader.someone', 'Someone') : '';
    if (user) {
        if (teammateNameDisplay === Preferences.DISPLAY_PREFER_NICKNAME) {
            name = user.nickname || getFullName(user);
        } else if (teammateNameDisplay === Preferences.DISPLAY_PREFER_FULL_NAME) {
            name = getFullName(user);
        } else {
            name = user.username;
        }

        if (!name || name.trim().length === 0) {
            name = user.username;
        }
    }

    return name;
}
