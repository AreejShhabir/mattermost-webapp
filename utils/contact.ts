// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {LicenseLinks} from './constants';

const baseZendeskFormURL = 'https://support.mattermost.com/hc/en-us/requests/new';

export enum ZendeskSupportForm {
    SELF_HOSTED_SUPPORT_FORM = '11184911962004',
    CLOUD_SUPPORT_FORM = '11184929555092',
    MOBILE_SUPPORT_FORM = '11184464799380',
    BILLING_LICENSING_SALES_SUPPORT_FORM = '360000852671',
}

export enum ZendeskFormFieldIDs {
    CLOUD_WORKSPACE_URL = '5245314479252',
    SELF_HOSTED_ENVIRONMENT = '360026980452',
    BILLING_SALES_CATEGORY = '360031056451',
    EMAIL = 'anonymous_requester_email',
    SUBJECT = 'subject',
}

export enum ZendeskFormFieldValues {
    SALES_GENERAL = 'sales_general',
    SALES_RENEWALS = 'sales-renewals',
    SALES_PRICING = 'sales_pricing',
}

export enum ZendeskFormPrefillSubjects {
    GENERAL_INQUIRY = 'General Inquiry',
    PASSWORD_RESET = 'Support for password reset process',
    LICENSE_PURCHASE = 'License purchase',
    LICENSE_RENEWAL = 'License renewal',
    LICENSE_EXPANSION = 'License expansion',
}

export type PrefillFieldFormFieldIDs = {
    id: ZendeskFormFieldIDs;
    val: string;
}

export const buildZendeskSupportForm = (form: ZendeskSupportForm, formFieldIDs: PrefillFieldFormFieldIDs[]): string => {
    let formUrl = `${baseZendeskFormURL}?ticket_form_id=${form}`;

    formFieldIDs.forEach((formPrefill) => {
        formUrl = formUrl.concat(`&tf_${formPrefill.id}=${formPrefill.val}`);
    });

    if (form === ZendeskSupportForm.SELF_HOSTED_SUPPORT_FORM) {
        formUrl = formUrl.concat(`&tf_${ZendeskFormFieldIDs.SELF_HOSTED_ENVIRONMENT}=production`);
    }

    return formUrl;
};

export const goToBillingGeneralSalesSupportForm = (email: string, subject: string) => {
    const form = ZendeskSupportForm.BILLING_LICENSING_SALES_SUPPORT_FORM;
    const url = buildZendeskSupportForm(form, [
        {id: ZendeskFormFieldIDs.EMAIL, val: email},
        {id: ZendeskFormFieldIDs.SUBJECT, val: subject},
        {id: ZendeskFormFieldIDs.BILLING_SALES_CATEGORY, val: ZendeskFormFieldValues.SALES_GENERAL},
    ]);
    window.open(url, '_blank');
};

export const goToSelfHostedSupportForm = (email: string, subject: string) => {
    const form = ZendeskSupportForm.SELF_HOSTED_SUPPORT_FORM;
    const url = buildZendeskSupportForm(form, [
        {id: ZendeskFormFieldIDs.EMAIL, val: email},
        {id: ZendeskFormFieldIDs.SUBJECT, val: subject},
    ]);
    window.open(url, '_blank');
};

export const goToCloudSupportForm = (email: string, subject: string, workspaceURL: string) => {
    const form = ZendeskSupportForm.SELF_HOSTED_SUPPORT_FORM;
    let url = buildZendeskSupportForm(form, [
        {id: ZendeskFormFieldIDs.EMAIL, val: email},
        {id: ZendeskFormFieldIDs.SUBJECT, val: subject},
    ]);
    url = url.concat(`&tf_${ZendeskFormFieldIDs.CLOUD_WORKSPACE_URL}=${workspaceURL}`);
    window.open(url, '_blank');
};

const encodeString = (s: string) => {
    return Buffer.from(s).toString('base64');
};

const buildMMURL = (baseURL: string, firstName: string, lastName: string, companyName: string, businessEmail: string) => {
    const mmURL = `${baseURL}?qk=${encodeString(firstName)}&qp=${encodeString(lastName)}&qw=${encodeString(companyName)}&qx=${encodeString(businessEmail)}`;
    return mmURL;
};

export const goToMattermostContactSalesForm = (firstName: string, lastName: string, companyName: string, businessEmail: string) => {
    const url = buildMMURL(LicenseLinks.CONTACT_SALES, firstName, lastName, companyName, businessEmail);
    window.open(url, '_blank');
};
