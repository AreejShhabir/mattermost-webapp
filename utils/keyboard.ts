// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {useEffect, useState} from 'react';

export const useGlobalKeyPressed = (key: string | ((e: KeyboardEvent) => boolean)) => {
    const [pressed, setPressed] = useState(false);

    const handleKeyDown = (e: KeyboardEvent) => {
        if (typeof key === 'string' ? e.key === key : key(e)) {
            setPressed(true);
        }
    };

    const handleKeyUp = (e: KeyboardEvent) => {
        if (typeof key === 'string' ? e.key === key : key(e)) {
            setPressed(false);
        }
    };

    useEffect(() => {
        window.addEventListener('keydown', handleKeyDown);
        window.addEventListener('keyup', handleKeyUp);

        return () => {
            window.removeEventListener('keydown', handleKeyDown);
            window.removeEventListener('keyup', handleKeyUp);
        };
    }, [key]);

    return pressed;
};
