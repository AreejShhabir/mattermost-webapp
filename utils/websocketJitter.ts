// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import random from 'lodash/random';

import {WebSocketMessage} from '@mattermost/client';
import {ActionFunc, DispatchFunc} from 'mattermost-redux/types/actions';

const JITTER_RANGE = 2000; // 2 seconds

export function wrapEventWithJitter(
    eventHandler: (websocketMessage: WebSocketMessage) => ActionFunc,
    websocketMessage: WebSocketMessage,
    dispatch: DispatchFunc | null = null,
) {
    const jitter = random(0, JITTER_RANGE);

    setTimeout(() => {
        if (dispatch) {
            // This is how all websocket events should be dispatched, and not from store directly
            // If you are writing a new websocket event, please use this
            dispatch(eventHandler(websocketMessage));
        } else {
            eventHandler(websocketMessage);
        }
    }, jitter);
}
