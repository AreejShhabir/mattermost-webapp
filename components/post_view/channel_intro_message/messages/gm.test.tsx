// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {boardComponent, channel, groupChannel, users} from './utils';

import GMIntroMessage from './gm';

describe('components/post_view/ChannelIntroMessages', () => {
    const baseProps = {
        currentUserId: 'test-user-id',
        channel,
        profiles: [],
    };

    describe('test Group Channel', () => {
        const props = {
            ...baseProps,
            channel: groupChannel,
        };

        test('should match snapshot, no profiles', () => {
            const wrapper = shallow(
                <GMIntroMessage
                    {...props}
                />,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('should match snapshot, with profiles, without boards', () => {
            const wrapper = shallow(
                <GMIntroMessage
                    {...props}
                    profiles={users}
                />,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('should match snapshot, with profiles, with boards', () => {
            const wrapper = shallow(
                <GMIntroMessage
                    {...props}
                    profiles={users}
                    boardComponent={boardComponent}
                />,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
});
