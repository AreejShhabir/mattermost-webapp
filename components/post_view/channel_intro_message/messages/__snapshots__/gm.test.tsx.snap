// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/post_view/ChannelIntroMessages test Group Channel should match snapshot, no profiles 1`] = `
<p
  className="channel-intro-text"
>
  <MemoizedFormattedMessage
    defaultMessage="This is the start of your group message history with these teammates. Messages and files shared here are not shown to people outside this area."
    id="intro_messages.group_message"
  />
</p>
`;

exports[`components/post_view/ChannelIntroMessages test Group Channel should match snapshot, with profiles, with boards 1`] = `
<Fragment>
  <div
    className="post-profile-img__container channel-intro-img"
  >
    <ProfilePicture
      hasMention={false}
      isEmoji={false}
      isRHS={false}
      key="introprofilepictureuser1"
      popoverPlacement="right"
      size="xl"
      src="/api/v4/users/user1/image?_=0"
      userId="user1"
      wrapperClass=""
    />
    <ProfilePicture
      hasMention={false}
      isEmoji={false}
      isRHS={false}
      key="introprofilepictureguest1"
      popoverPlacement="right"
      size="xl"
      src="/api/v4/users/guest1/image?_=0"
      userId="guest1"
      wrapperClass=""
    />
  </div>
  <p
    className="channel-intro-text"
  >
    <injectIntl(FormattedMarkdownMessage)
      defaultMessage="This is the start of your group message history with {names}.\\\\nMessages and files shared here are not shown to people outside this area."
      id="intro_messages.GM"
      values={
        Object {
          "names": "test channel",
        }
      }
    />
  </p>
  <Memo(BoardsButton)
    boardComponent={
      Object {
        "id": "board",
        "pluginId": "board",
      }
    }
    channel={
      Object {
        "create_at": 1508265709607,
        "creator_id": "creator_id",
        "delete_at": 0,
        "display_name": "test channel",
        "header": "test",
        "id": "channel_id",
        "last_post_at": 1508265709635,
        "name": "testing",
        "purpose": "test",
        "team_id": "team-id",
        "type": "G",
        "update_at": 1508265709607,
      }
    }
  />
  <Memo(SetHeaderButton)
    channel={
      Object {
        "create_at": 1508265709607,
        "creator_id": "creator_id",
        "delete_at": 0,
        "display_name": "test channel",
        "header": "test",
        "id": "channel_id",
        "last_post_at": 1508265709635,
        "name": "testing",
        "purpose": "test",
        "team_id": "team-id",
        "type": "G",
        "update_at": 1508265709607,
      }
    }
  />
</Fragment>
`;

exports[`components/post_view/ChannelIntroMessages test Group Channel should match snapshot, with profiles, without boards 1`] = `
<Fragment>
  <div
    className="post-profile-img__container channel-intro-img"
  >
    <ProfilePicture
      hasMention={false}
      isEmoji={false}
      isRHS={false}
      key="introprofilepictureuser1"
      popoverPlacement="right"
      size="xl"
      src="/api/v4/users/user1/image?_=0"
      userId="user1"
      wrapperClass=""
    />
    <ProfilePicture
      hasMention={false}
      isEmoji={false}
      isRHS={false}
      key="introprofilepictureguest1"
      popoverPlacement="right"
      size="xl"
      src="/api/v4/users/guest1/image?_=0"
      userId="guest1"
      wrapperClass=""
    />
  </div>
  <p
    className="channel-intro-text"
  >
    <injectIntl(FormattedMarkdownMessage)
      defaultMessage="This is the start of your group message history with {names}.\\\\nMessages and files shared here are not shown to people outside this area."
      id="intro_messages.GM"
      values={
        Object {
          "names": "test channel",
        }
      }
    />
  </p>
  <Memo(BoardsButton)
    channel={
      Object {
        "create_at": 1508265709607,
        "creator_id": "creator_id",
        "delete_at": 0,
        "display_name": "test channel",
        "header": "test",
        "id": "channel_id",
        "last_post_at": 1508265709635,
        "name": "testing",
        "purpose": "test",
        "team_id": "team-id",
        "type": "G",
        "update_at": 1508265709607,
      }
    }
  />
  <Memo(SetHeaderButton)
    channel={
      Object {
        "create_at": 1508265709607,
        "creator_id": "creator_id",
        "delete_at": 0,
        "display_name": "test channel",
        "header": "test",
        "id": "channel_id",
        "last_post_at": 1508265709635,
        "name": "testing",
        "purpose": "test",
        "team_id": "team-id",
        "type": "G",
        "update_at": 1508265709607,
      }
    }
  />
</Fragment>
`;
